RecoJets/JetProducers/src/PileupJetIdAlgo.cc

```diff
@@ -139,52 +140,42 @@ void setPtEtaPhi(const reco::Candidate & p, float & pt, float & eta, float &phi
 	phi = p.phi();
 }
 
-// ------------------------------------------------------------------------------------------
 void PileupJetIdAlgo::bookReader()
 {
 	if(etaBinnedWeights_){
           for(int v=0; v<nEtaBins_;v++){
-                std::unique_ptr<TMVA::Reader> etaReader = std::unique_ptr<TMVA::Reader>(new TMVA::Reader("!Color:Silent"));
-                etaReader_.push_back(std::move(etaReader));
-          }
-	} else {
-		reader_ = std::unique_ptr<TMVA::Reader>(new TMVA::Reader("!Color:Silent"));
-	}
-	if(etaBinnedWeights_){
-          for(int v=0; v<nEtaBins_;v++){
+            TMVA::Reader tmpTMVAReader( "!Color:Silent:!Error" );
             for(std::vector<std::string>::iterator it=tmvaEtaVariables_.at(v).begin(); it!=tmvaEtaVariables_.at(v).end(); ++it) {
-                if(  tmvaNames_[*it].empty() ) { 
-			tmvaNames_[*it] = *it;
-		}
-                etaReader_.at(v)->AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
             }
+            for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
+            }
+            reco::details::loadTMVAWeights(&tmpTMVAReader,  tmvaMethod_.c_str(), tmvaEtaWeights_.at(v).c_str() );
+            etaReader_.push_back(std::unique_ptr<const GBRForest> ( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmpTMVAReader.FindMVA(tmvaMethod_.c_str()) ) ) ) );
```

Comment:
Hi, since the producer that holds this algorithm is edm::stream it gets replicated for each thread (by thread I mean a thread on a processor) for CMSSW. Therefore if you run with four threads you have to build this algorithm four times. As it is implemented presently that means reading the MVA file and created four instances of the MVA which can blow up memory and start up time.

Using the global cache solves this problem permanently by only ever needing to build the MVAs once.